### PR TITLE
New coding contract: Shortest Path in a Grid

### PR DESCRIPTION
**Pull request by Ornedan**
Add a new pathfinding type coding contract: "Shortest Path in a Grid". Uses similar obstacle'd grid terrain as "Unique Paths in a Grid II", but permits movement in all cardinal directions.

localhost_8000__noscripts

### CODE REVIEW

File: src/data/codingcontracttypes.ts [hunk 1]
```diff
@@ -794,6 +795,135 @@ export const codingContractTypesMetadata: ICodingContractTypeMetadata[] = [
...
+      const dstX = width - 1;
+      const minPathLength = dstY + dstX; // Math.abs(dstY - srcY) + Math.abs(dstX - srcX)
+
+      const grid: number[][] = new Array(height);
+      for(let y = 0; y < height; y++)
+        grid[y] = new Array(width).fill(0);
+
+      for(let y = 0; y < height; y++) {
+        for(let x = 0; x < width; x++) {
+          if(y == 0 && x == 0) continue; // Don't block start
```
Comment (phyzical) [inline src/data/codingcontracttypes.ts:838]: we could consider allowing the start and end being blockable and just expect "" also, more edges for the dev to handle and come across at a random time

**Review by phyzical — APPROVED** [review]
looks resonable to me, though i didnt deep dive into the heap. but the solvers looks legit, check length and check answer doesn't hit an obstacle and make sure it ends at the end 👍

File: src/data/codingcontracttypes.ts (hunk 1 above)
Comment (Ornedan) [reply src/data/codingcontracttypes.ts:838]: The problem starts with the "character" standing at the (0,0) grid location, which requires that location to not be blocked.

Blocking the end could be OK, but if it is blockable then the obstacle generation weights would also need to be adjusted. Just removing the "Don't block destination" special case drops solvable puzzles ratio to ~66%.

Comment (phyzical) [reply src/data/codingcontracttypes.ts:838]: no worries :)

**Comment by Hedrauta** [comment]
🤔 what if a pattern allows multiple answers?
Maybe something like:
```
[
[0,1,0,0,0,0],
[0,0,0,0,1,0]]
```` 
Then, DRRRURRD or DRRURRRD  would be correct 🤔 
will the solver fetch all possible answers and chose the shortest ones for checking the user's answer, or does it only allow one?

**Comment by phyzical** [comment]
when i read the source i think we are enforcing is it the same length as the first found answer, is what they provided a valid answer and does it end in the last corner.

so if there is multiple shortest paths it will allow any

**Comment by Ornedan** [comment]
Yes. Solution checker process:
- find **a** shortest path
- require that the player-given answer is not longer than the found path
- require that the player-given path is a valid path (stays in bounds, does not attempt to move into obstacles)
- require that following the player-given path ends up at the target coordinates 

If there are multiple shortest paths, any one will be accepted because they all fill all requirements.

**Pull request merged**